### DISCUSSION TITLE
Make `rich.traceback` capable of handling `Exception` objects which are `RichRenderable`

### DISCUSSION THREAD

**Discussion by pradyunsg** [Textualize/rich#2717] [Ideas]
In a few of the projects I've been working on (pip, sphinx-theme-builder and more), I have been creating "rich-aware" errors -- i.e. these error objects are `RichRenderable`, by implementing the `__rich_console__`/`__rich__` methods.

It would be really nice if `rich.traceback` were to use the "rich" representation of those errors (falling back to `<error rendering {exc.__class__.__name__}>`, like it does with `str` calls currently).

This would make it possible to switch to using Rich's traceback handler exclusively, allowing for nicer tracebacks for users while still presenting the `RichRenderable` exception objects in a pretty manner, instead of degrading those to plain strings. :)

**Comment by willmcgugan** [comment]
So you make the exceptions themselves renderable? I do something similar in Textual.

I guess you always want the traceback, but with the exception rendered if it is renderable?

**Reply by pradyunsg** [reply]
> So you make the exceptions themselves renderable?

Yes. https://github.com/pradyunsg/sphinx-theme-builder/blob/b43870f8db9a47e3c81f2244bb0e8ec0b2a33f3a/src/sphinx_theme_builder/_internal/errors.py#L80

> I guess you always want the traceback, but with the exception rendered if it is renderable?

Yes. #2441 would be nice as well, but we'd chatted briefly about that back when pip added rich into it.

**Reply by willmcgugan** [reply]
Up for a PR? I'm unlikely to get to that in the short term.

I'd also accept a PR for a no-panel switch on Traceback. Probably not a "empty" box though, as that will add whitespace and I'll get folk complaining about that.

**Reply by pradyunsg** [reply]
I'll give it a shot today. If I don't file a PR in the next ~24 hours, consider this up for grabs. :)

**Reply by pradyunsg** [reply]
(FWIW, I've reported the other message as spam, via GitHub support)

**Comment by pradyunsg** [comment]
OK, I finally came around to this. https://github.com/Textualize/rich/pull/3325 is the PR, and it works!

**Comment by pradyunsg** [comment]
Also I've filed https://github.com/Textualize/rich/pull/3324 -- a somewhat-unrelated PR that cleans up some dead code from traceback tests.
